## 1) Even Numbers using for loop

In [1]:
n = int(input("Enter a number: "))

for i in range(1, n + 1):
    if i % 2 == 0:
        print(i, "is Even")

Enter a number:  8


2 is Even
4 is Even
6 is Even
8 is Even


In [2]:
n = int(input("Enter a number: "))

if n % 2 == 0:
    print("Even Number")
else:
    print("Not Even")

Enter a number:  3


Not Even


In [7]:
for i in range(1, 11):
    if i % 2 ==0:
        print(i)

2
4
6
8
10


In [14]:
n = int(input())
for i in range(n):
    if i % 2==0:
        print(i, "is even")

 5


0 is even
2 is even
4 is even


### 2) Even Numbers using while loop

In [21]:
n = int(input())
i = 1
while i < n:
    if i % 2==0:
        print(i, "is even")
    i+=1  

 5


2 is even
4 is even


### 3) Prime Numbers using for loop

In [3]:
n = int(input())
for i in range(2, n):
    if n% i ==0:
        print("not Prime number")
        break
else:
        print("prime")   

 8


not Prime number


In [9]:
n = int(input())
if n<=1:
    print("not a prime number")
else:
    for i in range(2, n):
        if n % i==0:
            print("not a prime number")
            break
    else:
        print("prime numebr")

 3


prime numebr


### 4) Prime Numbers using while loop

In [24]:
n = int(input())
i = 2
while i <=n:
    if n % i==0:
        print("not prime")
        break
    i+=1
else:
    print("prime") 

 2


not prime


## 5) Write Python functions to print even and prime numbers from 1 to 20.

In [26]:
def print_even():
    for i in range(1, 5):
        if i % 2 == 0:
            print(i)

def print_prime():
    for num in range(1, 4):
        if num > 1:
            for i in range(2, num):
                if num % i == 0:
                    break
            else:
                print(num)

print("Even Numbers:")
print_even()

print("Prime Numbers:")
print_prime()

Even Numbers:
2
4
Prime Numbers:
2
3


### 6) Print numbers from 1 to 10

In [28]:
for i in range(1, 11):
    print(i)

1
2
3
4
5
6
7
8
9
10


### 7) Print all odd numbers from 1 to 20

In [36]:
n = 21
for i in range(1, n):
    if i % 2 !=0:
        print(i, "Odd")

1 Odd
3 Odd
5 Odd
7 Odd
9 Odd
11 Odd
13 Odd
15 Odd
17 Odd
19 Odd


### 8) Print the table of a number (like 5 x 1 = 5 ...)

In [65]:
n = int(input())
for i in range(1, 11):
    print(n, "x", i, "=", n*i)

 4


4 x 1 = 4
4 x 2 = 8
4 x 3 = 12
4 x 4 = 16
4 x 5 = 20
4 x 6 = 24
4 x 7 = 28
4 x 8 = 32
4 x 9 = 36
4 x 10 = 40


In [1]:
n = int(input())
for i in range(n):
    if i % 2 !=0:
        print(i)

 10


1
3
5
7
9


In [75]:
n = int(input())
for i in range(1, 4):
    for j in range(1, 11):
        print(i, "x", j, "=", i*j)
    print()

 3


1 x 1 = 1
1 x 2 = 2
1 x 3 = 3
1 x 4 = 4
1 x 5 = 5
1 x 6 = 6
1 x 7 = 7
1 x 8 = 8
1 x 9 = 9
1 x 10 = 10

2 x 1 = 2
2 x 2 = 4
2 x 3 = 6
2 x 4 = 8
2 x 5 = 10
2 x 6 = 12
2 x 7 = 14
2 x 8 = 16
2 x 9 = 18
2 x 10 = 20

3 x 1 = 3
3 x 2 = 6
3 x 3 = 9
3 x 4 = 12
3 x 5 = 15
3 x 6 = 18
3 x 7 = 21
3 x 8 = 24
3 x 9 = 27
3 x 10 = 30



### 9) Print the sum of numbers from 1 to 100

In [49]:
n = 101
sum =0
for i in range(1, n):
    sum+=i
print(sum)

5050


### 10) Print squares of numbers from 1 to 10

In [51]:
for i in range(1, 11):
    print(i*i)

1
4
9
16
25
36
49
64
81
100


### 11) Print a pattern like this:

In [63]:
for i in range(1, 5):
    print(" * " * i)

 * 
 *  * 
 *  *  * 
 *  *  *  * 


### 12) Count how many even numbers are in a list

In [71]:
lst = [1, 2, 3, 4, 5, 6, 8, 10]
count = 0
for i in lst:
    if i % 2==0:
        count+=1
print(count)        

5


### 13) Find the largest number in a list using a for loop

In [77]:
lst = [1, 2, 3, 4, 59, 100, 2000000, 90]
max_num = lst[0]
for i in lst:
    if i > max_num:
        max_num = i
print(max_num)  

2000000


### 14) Print reverse numbers from 10 to 1

In [82]:
n = 10
for i in range(n, 0, -1):
    print(i)

10
9
8
7
6
5
4
3
2
1


### 15) Print all characters in a string one by one

In [85]:
a = "sindhu"
for i in a:
    print(i)

s
i
n
d
h
u


## FUNCTION-BASED QUESTIONS

### 1) Write a function to return the square of a number

In [4]:
def square(n):
    a = n*n
    return a
square(2)

4

### 2) Write a function to check if a number is even or not

In [90]:
def even(num):
    if num %2==0:
        print("even")
    else:
        print("not even")
even(7)

not even


### 3) Write a function to check if a number is prime

In [100]:
def prime(n):
    i = 2
    for i in range(2, n):
        if n % i==0:
            print("not prime")
            break
    else:
        print("prime")
prime(8)

not prime


### 4) Write a function to find the factorial of a number

In [103]:
def factorial(n):
    fact = 1
    for i in range(1, n+1):
        fact = fact*i
    return fact
factorial(5)

120

### 5) Write a function to count vowels in a string

In [110]:
def vowel(x):
    count = 0
    for i in x.lower():
        if i in "aeiou":
            count+=1
    return count
vowel("I am Sindhu")   

4

### 6) Write a function that takes a list and returns the sum

In [112]:
def sum1(lst):
    sum2 = 0
    for i in lst:
        sum2+= i
    return sum2
sum1([1, 2, 3, 4, 5])

15

### 7) Write a function that takes a list and returns only even numbers

In [118]:
def even(lst):
    evens = []
    for i in lst:
        if i % 2==0:
            evens.append(i)
    return evens
even([1, 2, 3, 8, 9])

[2, 8]

### 8) Write a function to reverse a string

In [125]:
def reverse1(x): 
    for i in x:
        c= x[::-1]
        return c 
reverse1("sindhu")  

'uhdnis'

In [126]:
def reverse_string(x):
    reversed_str = ""
    for char in x:
        reversed_str = char + reversed_str
    return reversed_str

print(reverse_string("sindhu"))

uhdnis


In [127]:
text = input("Enter the string: ")
result = ""

for i in range(0, len(text), 2):
    char = text[i]
    count = int(text[i+1])
    result += char * count

print(result)

Enter the string:  a2b3c4


aabbbcccc


## input: 1789
### output: 72

In [191]:
n = 1789
n = str(n)
max_mul = 0

for i in range(len(n) - 1):
    mul = int(n[i]) * int(n[i + 1])
    if mul > max_mul:
        max_mul = mul

print(max_mul)

72


### 9) Max Sum of Two Adjacent Digits
### Input: 7345 → Output: 7 + 3 = 10, 3 + 4 = 7, 4 + 5 = 9 → Max sum is 10
### Output: 10

In [139]:
n = 7345
n = str(n)
max_sum = 0
for i in range(len(n)-1):
    sum = int(n[i]) + int(n[i+1])
    if sum > max_sum:
        max_sum = sum
print(max_sum)    

10


### 10) Max Difference Between Two Adjacent Digits
### Input: 4289 → Output: |4-2|=2, |2-8|=6, |8-9|=1 → Max diff is 6
### Output: 6

In [143]:
n = 4289
n = str(n)
max_dif = 0
for i in range(len(n)-1):
    dif = abs(int(n[i]) - int(n[i+1]))
    if dif > max_dif:
        max_dif = dif
print(max_dif)

6


### 11) Sum of Digits at Even Indexes
#### Input: 123456 → Indexes 0,2,4 → 1+3+5 = 9
#### Output: 9

In [152]:
n = 123456
n = str(n)
sum_even_idx = 0
for i in range(0, len(n), 2):  # Only even indexes
    sum_even_idx += int(n[i])
print(sum_even_idx)

9


### 12) Multiply All Digits Except Zero
#### Input: 1042 → 1 × 4 × 2 = 8
##### Output: 8

In [156]:
n = 1042
n = str(n)
product =1
for i in n:
    if i !='0':
        product *= int(i)
print(product)     

8


### 13) Count Digits Greater than Previous Digit
#### Input: 132578 → Compare each digit with the previous:
##### 3 > 1 ✅, 5 > 2 ✅, 7 > 5 ✅, 8 > 7 ✅ → Count = 4
##### Output: 4

In [169]:
n = 132578
n = str(n)
count = 0
for i in range(1, len(n)):
    if int(n[i]) > int(n[i-1]):
        count+= 1
print(count)

4


### 14) Find Digit Pair That Adds to 10
### Input: 58392 → 5+3=8 ❌, 8+3=11 ❌, 3+9=12 ❌, 9+2=11 ❌
#### → No such pair
##### Output: No pair

In [173]:
n = 58392
n = str(n)
digit_pair=0
for i in range(len(n)-1):
    if int(n[i]) + int(n[i+1]) == 10:
        digit_pair+=1
if digit_pair > 0:
    print("Total Pair", digit_pair)
else:
    print("No such Pair")      

No such Pair


## 1) Check if a given string is a palindrome.
### Input: "madam" → Output: True

In [177]:
n = "madam"
if n == n[::-1]:
    print("True")
else:
    print("False")

True


## 2) Check if a given number is a palindrome.
### Input: 121 → Output: True

In [179]:
n = 121
if str(n) == str(n)[::-1]:
    print("True")
else:
    print("False")

True


## 3) Count how many palindromes are in a given list of strings.
### Input: ["madam", "apple", "racecar", "level"] → Output: 3

In [183]:
n = ["madam", "apple", "racecar", "level"]
count = 0
for i in n:
    if i == i[::-1]:
        count+= 1
print(count)   

3


## 4) Check if a sentence is a palindrome (ignore spaces and punctuation).
### Input: "Was it a car or a cat I saw" → Output: True

In [33]:
n = "Was it a car or a cat I saw"

cleaned = ''.join(char.lower() for char in n if char.isalnum())

if cleaned == cleaned[::-1]:
    print(cleaned)
else:
    print("False")

wasitacaroracatisaw


# Armstrong Number:

In [36]:
num = int(input("Enter a number: "))
sum = 0
order = len(str(num))  # number of digits
for digit in str(num):
    sum += int(digit) ** order
if sum == num:
    print(num, "is an Armstrong number.")
else:
    print(num, "is not an Armstrong number.")

Enter a number:  153


153 is an Armstrong number.


In [2]:
num = int(input())
sum = 0
order = len(str(num))
for digit in str(num):
    sum+= int(digit) ** order
if sum == num:
    print(num, "is an Armstrong num.")
else:
    print(num, "is not an Armstrong num.")

 153


153 is an Armstrong num.


## twin prime numbers

In [23]:
for num in range(2, 15):
    # check if num is prime
    prime1 = True
    for i in range(2, num):
        if num % i == 0:
            prime1 = False
            break

    # check if num + 2 is also prime
    prime2 = True
    for j in range(2, num + 2):
        if (num + 2) % j == 0:
            prime2 = False
            break

    # if both are prime, print as twin prime
    if prime1 and prime2:
        print(num, "and", num + 2, "are twin primes")

3 and 5 are twin primes
5 and 7 are twin primes
11 and 13 are twin primes


In [25]:
n = "a3b4c4"
output = ""
for i in range(0, len(n), 2):
    char = n[i]
    count = int(n[i+1])
    output += char * count
print(output)

aaabbbbcccc


## a2#3c4

In [193]:
n= "a2#3c4"
output = ""
for i in range(0, len(n), 2):
    char = n[i]
    count = int(n[i+1])
    output+= char*count
print(output)

aa###cccc


In [2]:
def lst1(l):
    myout = []
    for i in l:
        if type(i) == list:
            for sub in i:
                if type(sub) == list:
                    for x in sub:
                        myout.append(x)
                else:
                    myout.append(sub)
        else:
            myout.append(i)
    return myout
l = [1, [2, 3], 4, [5, 6]]
result = lst1(l)
print(result)

[1, 2, 3, 4, 5, 6]


In [15]:
l = [1, [2, 3], 4, [5, 6, 7], 8]
int1 = 0
lst1= 0
for i in l:
    if type(i) == int:
        int1+=1
    elif type(i) == list:
        lst1+= 1
    else:
        print("unknown type") 
print(int1, "integers found")
print(lst1, "list found")

3 integers found
2 list found


## Count Types in List (int, list, str)
### l = [1, [2, 3], 'hello', 5, [6, 7], 'hi', 8]


In [16]:
l = [1, [2, 3], 'hello', 5, [6,7], 'hi', 8]
int_count = 0
list_count =0
str_count=0
for i in l:
    if type(i) == int:
        int_count+=1
    elif type(i) == list:
        list_count+=1
    elif type(i) == str:
        str_count+=1
    else:
        print("not found")
print("integers are", int_count)
print("list are", list_count)
print("strings are", str_count)

integers are 3
list are 2
strings are 2


## Print Elements from Only Lists
### l = [1, [2, 3], 4, [5, 6], 7]

In [21]:
l = [1, [2, 3], 4, [5, 6], 7]
for i in l:
    if type(i) == list:
        for x in i:
            print(x)

2
3
5
6


In [23]:
l = [2, 5, 7, 10, 12, 15, 18]
even_count=0
odd_num=0
for i in l:
    if i % 2==0:
        even_count+=1  
    else:
        odd_num+=1
print("even numbers:", even_count)
print("odd numbers", odd_num)

even numbers: 4
odd numbers 3


## Find the Largest Digit in a Number

In [38]:
n = "54823"
max_digit=0
for i in n:
    if int(i) > max_digit:
        max_digit =int(i)
print(max_digit)

8


In [1]:
l = [1, [2, 3], 6, [89, 98], [34, 23]]
myout = []
for i in l:
    if type(i) == list:
        for j in i:
            myout.append(j)
    else:
        myout.append(i)
print(myout)

[1, 2, 3, 6, 89, 98, 34, 23]


In [8]:
l1=[1, 2, 3, 3]
sum=0
for i in l1:
    sum+=i
print(sum)

9


In [2]:
s = input()  

roman_map = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
total = 0
prev = 0

for char in reversed(s):
    value = roman_map[char]
    if value < prev:
        total -= value
    else:
        total += value
    prev = value

print(total)  

 M


1000


In [76]:
nums = eval(input())  
sum = 0

for i in nums:
    if nums.count(i) == 1:  
        sum += i            
print(sum)  

 [1, 3, 3, 8, 8]


1


In [14]:
n = eval(input())         # Read the list
for i in range(1, len(n)+1): # Go from 0 to n (inclusive)
    if i not in n:        # If number is missing in list
        print(i)          # Print missing number
        break             # Stop after finding it

 [1, 3]


2


In [18]:
n = eval(input())         # Read the list
for i in range(len(n)+1): # Go from 0 to n (inclusive)
    if i not in n:        # If number is missing in list
        print(i)          # Print missing number
        break     

 [4, 3, 1]


0


In [22]:
n = int(input())

# Base cases
if n == 1:
    print(1)
elif n == 2:
    print(2)
else:
    a, b = 1, 2
    for _ in range(3, n + 1):
        a, b = b, a + b
    print(b)

 3


3


## Armstrong Number Check

In [24]:
num = int(input("Enter a number: "))
sum = 0
temp = num

while temp > 0:
    digit = temp % 10
    sum += digit ** 3
    temp = temp // 10

if sum == num:
    print("Armstrong number")
else:
    print("Not an Armstrong number")

Enter a number:  123


Not an Armstrong number


In [3]:
num = int(input("Enter a number: "))
temp = num
num_digits = len(str(num))
sum = 0

for digit in str(num):
    sum += int(digit) ** num_digits  # Raise each digit to power of number of digits

if sum == temp:
    print("Armstrong number")
else:
    print("Not an Armstrong number")

Enter a number:  9474


Armstrong number


##  Use a for loop to print Fibonacci series up to n terms.

In [25]:
n = int(input("Enter number of terms: "))
a = 0
b = 1
for i in range(n):
    print(a, end=' ')
    c = a + b
    a = b
    b = c

Enter number of terms:  20


0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 

In [11]:
text = "python is a general purpose language"
word = input("Enter your search : ")
start = text.find(word)
if start != -1:
    end = start + len(word) - 1
    print(f"The word {word} is starting from {start} to {end}")
else:
    print("Word not found in text")

Enter your search :  general


The word general is starting from 12 to 18


## Reversed string

In [3]:
text= input()
reversed_str= " "
for i in text:
    reversed_str = i+reversed_str
print(reversed_str)

 prashanth


htnahsarp 


## find factorial

In [7]:
num = int(input())
fact = 1
for i in range(1, num+1):
    fact = fact*i
print(fact)

 5


120


## Count Vowels in a String

In [14]:
str1= input("Enter a string: ")
count=0
for i in str1.lower():
    if i in "aeiou":
        count+=1
print(count)

Enter a string:  sindhu


2


## Sum of Digits
### Input: 1234
### Output: 1 + 2 + 3 + 4 = 10


In [32]:
num = int(input())
count=0
for i in str(num):
    count+= int(i)
print(count)

 1234


10


## Check for Palindrome 

In [50]:
str1 = "madam"
reversed_str = ""
for i in str1:
    reversed_str = i + reversed_str  
if str1 == reversed_str:
    print("Yes, palindrome")
else:
    print("Not a palindrome")

Yes, palindrome


## Print All Unique Letters in a String (No Repeats)
### Input: "sindhu"
### Output: s i n d h u

### Input: "apple"
### Output: a e

In [60]:
str1 = "apple"
for i in str1:
    if str1.count(i) == 1:
        print(i, end=" ")      

a l e 

In [58]:
str1 = "sindhu"
for i in str1:
    if str1.count(i) == 1:
        print(i, end=" ")      

s i n d h u 

## Find the Second Largest Number in a List
### Input: [5, 1, 8, 7, 3]
### Output: 7



In [62]:
lst1 = [5, 1, 8, 7, 3]
max_lar=lst1[0]
for i in lst1:
    if i> max_lar:
        max_lar = i
second_max =lst1[0]
for i in lst1:
    if i != max_lar and i>second_max:
        second_max=i
print(second_max)       

7


## Count Digits, Letters, and Others in a String
### Input: "Sindhu123@!"
### Output: Letters: 6, Digits: 3, Others: 2

In [63]:
str1 = input("Enter a string: ")

letters = 0
digits = 0
others = 0

for char in str1:
    if char.isalpha():
        letters += 1
    elif char.isdigit():
        digits += 1
    else:
        others += 1

print("Letters:", letters)
print("Digits:", digits)
print("Others:", others)

Enter a string:  sindhu hi


Letters: 8
Digits: 0
Others: 1


### Given a list of integers, write a program to find the number closest to 0.
### If there are multiple numbers equally close to 0, return the larger number.
### The input will be a list like [-4, -2, 1, 4, 8].
### The output will be a single integer that is closest to 0 based on the rules.

In [12]:
nums = [-4, -2, 1, 4, 8]
closest = nums[0]
for n in nums:
    if abs(n) < abs(closest):
        closest = n
    elif abs(n) == abs(closest) and n > closest:
        closest = n
print(closest)

1


## remove all duplicate characters in a string

In [4]:
s = input()    # Take input string
result = ""    # To store final result

for char in s:
    if char not in result:   # Add only if not already in result
        result += char

print(result)  # Output without duplicates, in order

 innomatics


inomatcs


### Each digit from the number is alternately added and subtracted, starting from positive for the first digit.

In [5]:
n = input()      # Get number as a string
result = 0       # Start with 0
i = 0
while i < len(n):
    digit = int(n[i])   # Take one digit at a time
    if i % 2 == 0:      # If index is even, add
        result += digit
    else:               # If index is odd, subtract
        result -= digit
    i += 1              # Move to next digit
print(result)           # Print the final answer

 521


4


In [4]:
n = input()      # Take number as string
result = 0
for i in range(len(n)):
    digit = int(n[i])   # Get each digit
    if i % 2 == 0:      # Even index → add
        result += digit
    else:               # Odd index → subtract
        result -= digit
print(result)           # Final answer

 543


4


## remove all duplicate characters in a string

In [41]:
str1="innomatics"
str2 = " "
for i in str1:
    if i not in str2:
        str2 +=i
print(str2)

 inomatcs


In [5]:
n = 5
for i in range(0, n):
    print(i*i)

0
1
4
9
16


In [2]:
ones = ["", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
tens = ["", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]
teens = ["ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", 
         "sixteen", "seventeen", "eighteen", "nineteen"]

n = int(input("Enter number (0–9999): "))
words = ""

if n == 0:
    words = "zero"
else:
    if n >= 1000:
        digit = n // 1000
        words += ones[digit] + " thousand "
        n = n % 1000

    if n >= 100:
        digit = n // 100
        words += ones[digit] + " hundred "
        n = n % 100

    if n >= 20:
        digit = n // 10
        words += tens[digit] + " "
        n = n % 10

    elif n >= 10:
        words += teens[n - 10] + " "
        n = 0

    if n > 0:
        words += ones[n] + " "

print(words.strip())

Enter number (0–9999):  1000


one thousand


## Question 1: Alternating Sum of Digits (Right to Left)
#### Example:
- Input: 1234
- Output: 4 - 3 + 2 - 1 = 2



In [12]:
n = 1234
n= str(n)
sum=0
for i in range(len(n)):
    digit = int(n[len(n) - 1 - i])
    if i%2==0:
        sum+=digit
    else:
        sum-=digit
print(sum)

2


###  Question 2: Count Digits Greater Than Previous Digit
- Logic: For each digit in the number, count how many digits are greater than the digit before it.

- Example:
- Input: 13542
- Output: 2 (3 > 1 and 5 > 3)



In [14]:
n = 13542
n=str(n)
count =0
for i in range(1, len(n)-1):
    prev = int(n[i-1])
    curr= int(n[i])
    if curr > prev:
        count+=1
print(count)

2


### Question 3: Swap Even and Odd Indexed Digits
- Logic: Swap every even and odd index digit pair in the number.

- Example:
- Input: 123456
- Output: 214365
- Explanation: Swap 1↔2, 3↔4, 5↔6

In [16]:
n = input()
n= list(n)
for i in range(0, len(n)-1, 2):
    n[i], n[i+1] = n[i+1], n[i]
print("".join(n))

 123456


214365


In [29]:
n = "1234567"
n = list(n)  
for i in range(0, len(n)-1, 2):  
    n[i], n[i+1] = n[i+1], n[i] 
print("".join(n))  

2143657


In [15]:
c=0
n=input()
for i in range(len(n)-1):
    if n[i:i+1] > n[i+1:i+2]:
        c=c+1
print(c)

 13542


2


## Question 4: Find the Sum of Digits at Prime Indices
- Input: 1234567
- Digits at prime indices = 2, 3, 4, 6 → sum = 15

In [22]:
n = input()
sum = 0
for i in range(len(n)):
    if i > 1:
        is_prime = True
        for j in range(2, i):
            if i % j == 0:
                is_prime = False
                break
        if is_prime:
            sum += int(n[i])

print(sum)

 1234567


13


In [26]:
def prime(n):
    count=0
    for i in range(2,n+1):
        if n%i==0:
            count+=1
    if count==1:
        return True
    return False
ls=[1,2,3,4,5,6,7]
count=0
for i in range(len(ls)):
    if prime(i):
        count+=ls[i]
print(count)

13


In [27]:
n = "1234567"
sum = 0

for i in range(len(n)):
    if i > 1:  # Prime numbers start from index 2
        is_prime = True
        for j in range(2, i):
            if i % j == 0:
                is_prime = False
                break
        if is_prime:
            sum += int(n[i])  # Add digit at prime index

print(sum)

13


## Reverse Words in a Sentence(Without slicing)

In [12]:
n = "sindhu"
revrsed = ""
print(revrsed.join(reversed(n))) 

uhdnis


## Count Vowels in a String

In [13]:
n = "I am Sindhu"
count=0
for i in n.lower():
    if i in "aeiou":
        count+=1
print(count)

4


## Find the word with the maximum length in a sentence

In [21]:
n = input()
words = n.split()
max_word = " "
for i in words:
    if len(i) > len(max_word):
        max_word=i
print(max_word)

 I love programming in Python


programming


## Count digits in a given string

In [26]:
n = "Sindhu123"
count=0
for i in n:
    if i.isdigit():
        count+=1
print(count) 

3


## Count Consonants in a String
- Input: "Hello World"
- Expected Output: 7

In [30]:
n = "Hello World"
count = 0
for i in n.lower():
    if i.isalpha() and i not in "aeiou":
        count+=1
print(count)

7


## Check for Palindrome

In [36]:
n = "madam"
n = n.lower()
word=""
revrsed = word.join(reversed(n))
if n == revrsed:
    print("palindrome")
else:
    print("not palindrome")

palindrome


In [37]:
n = "sindhu"
n = n.lower()
word=""
revrsed = word.join(reversed(n))
if n == revrsed:
    print("palindrome")
else:
    print("not palindrome")

not palindrome


## Count how many words are in a sentence.
- Input: "I am learning Python programming"
- Expected Output: 5

In [38]:
n = "I am learning Python programming"
word = n.split()
count = 0
for i in word:
    if len(i) > 0:
        count+=1
print(count)

5


## Check if a number is a perfect square.
-  Input: 16
- Output: "Perfect square"
- Input: 18
- Output: "Not a perfect square"

In [56]:
n = 16
for i in range(1, n+1):
    if n == pow(i, 2):
        print("perfect suqare")   
        break
else:
    print("not perfect squre")

perfect suqare


In [57]:
n = 16
is_square = False

for i in range(1, n + 1):
    if i * i == n:
        is_square = True
        break

if is_square:
    print("Perfect square")
else:
    print("Not a perfect square")

Perfect square


## Find the sum of all even digits in a number.
- Input: n = 482917
- Expected Output: 4 + 8 + 2 = 14

In [59]:
n = input()
sum=0
for i in n:
    if int(i) % 2==0:
        sum+= int(i)
print(sum)

 482917


14


## Count how many words in a sentence start with a vowel.

-  Input: "An eagle and an owl are animals"
- Expected Output: 7

In [63]:
n = "An eagle and an owl are animals"
word = n.split()
count=0
for i in word:
    if i[0].lower() in "aeiou":
        count+=1
print(count)    

7


##  Count Words Ending with a Vowel

In [64]:
n = "Mangoes are ripe and tasty"
word =n.split()
count=0
for i in word:
    if i[-1].lower() in "aeiou":
        count+=1
print(count)

2


## Sum of Digits at Odd Index Positions
- Input:
- A number as a string (example: "123456").
- Task:
- Find the sum of digits at odd index positions (indexing starts from 0).

In [68]:
n = "67890"
sum_digits = 0
for i in range(len(n)):
    if i % 2 != 0:  # odd index
        sum_digits += int(n[i])        
print(sum_digits)

16


## Question: Count how many times 'a' appears in a string
- Input:
- banana

- Output:
- Count of 'a': 3

- Logic:
- Use string.count('a') to directly count the letter 'a'

In [1]:
n = "banana"
print(n.count('a'))

3


## Digit Product and Sum Check
- Input: 123
- Output:  (1+2+3 = 6, 1*2*3 = 6 → Equal)

- Input: 112
- Output: Not Equal  (1+1+2 = 4, 1*1*2 = 2)


In [8]:
n = input()
sum1 =0
product = 1
for i in n:
    digit = int(i)
    sum1+= digit
    product *=digit
if product == sum1:
    print("equal")
else:
    print("Not Equal")

 112


Not Equal


##  Alternating Sum of Digits
- Problem:
- Given a number, compute the alternating sum of its digits:
- digit1 - digit2 + digit3 - digit4 + ...
- (Start with addition at index 0)
- Input: 12345
- Output: 1 - 2 + 3 - 4 + 5 = 3

- Input: 246
- Output: 2 - 4 + 6 = 4


In [20]:
n = input()
sum1 =0
for i in range(len(n)):
    digit = int(n[i])
    if i % 2 == 0:
        sum1 +=digit
    else:
        sum1 -=digit
print(sum1)
        

 12345


3


##  Count of 1s in a Number
- Problem:
- Given a number, count how many times the digit 1 appears in it
- Input: 151241
- Output: 3

In [22]:
n = input()
count = 0
for i in n:
    if int(i) == 1:
        count+=1
print(count)

 151241


3


In [23]:
n = input()
print(n.count('1'))

 151241


3


## Input: 123321
#### Output: Balanced

## Input: 123456
### Output: Not Balanced


In [24]:
n = input("Enter a number: ")

length = len(n)
half = length // 2

first_half_sum = 0
second_half_sum = 0

for i in range(half):
    first_half_sum += int(n[i])

for i in range(half, length):
    second_half_sum += int(n[i])

if first_half_sum == second_half_sum:
    print("Balanced")
else:
    print("Not Balanced")

Enter a number:  123321


Balanced


## 

    "1. Write a function to check if a given string is a palindrome (ignore spaces and case).",
    "2. Write a function that returns the nth Fibonacci number using memoization.",
    "3. Create a function to return the longest substring without repeating characters.",
    "4. Write a function to determine if two strings are anagrams of each other.",
    "5. Write a function to find the majority element in a list (element appearing > n/2 times).",
    "6. Implement a function that rotates a 2D matrix 90 degrees clockwise.",
    "7. Write a function to find all prime numbers less than a given number n.",
    "8. Create a function that takes a list of numbers and returns all the subsets.",
    "9. Write a function to implement binary search recursively.",
    "10. Create a function that finds the missing number in a list of integers from 1 to n.",
    "11. Write a function that returns the sum of digits of a number until the sum is a single digit.",
    "12. Write a function that reverses the words in a sentence but keeps their order.",
    "13. Write a function that checks whether a number is a power of 2.",
    "14. Create a function to group a list of words into sets of anagrams.",
    "15. Write a function that returns the first non-repeating character in a string.",
    "16. Create a function to count the number of vowels and consonants in a string.",
    "17. Write a function to find the longest common prefix in a list of strings.",
    "18. Write a function that returns True if a string can be rearranged to form a palindrome.",
    "19. Create a function to merge two sorted lists into one sorted list without using sort().",
    "20. Write a recursive function to compute the factorial of a number.",
    "21. Implement a function that checks if a Sudoku board is valid (only row, col, and box constraints).",
    "22. Write a function to generate all permutations of a string.",
    "23. Write a function that returns the intersection of two lists (without using sets).",
    "24. Create a function that flattens a deeply nested list using recursion.",
    "25. Write a function that implements Caesar cipher encryption and decryption."


## 1) Write a function to check if a given string is a palindrome (ignore spaces and case).",

In [3]:
def palindrome(n):
    n =n.replace(" ", "").lower() 
    reverse = n[::-1]
    if n == reverse:
        print("Palidnrome")
    else:
        print("not palindrome")
palindrome("madam")
palindrome("sindhu")

Palidnrome
not palindrome


In [4]:
def palindrome(n):
    n =n.replace(" ", "").lower() 
    reverse = ""
    for i in n:
        reverse = i+reverse
    if n == reverse:
        print("Palidnrome")
    else:
        print("not palindrome")
palindrome("madam")
palindrome("sindhu")

Palidnrome
not palindrome


## 2. Write a function that returns the nth Fibonacci number using memoization.

In [6]:
def fibo(n, memo={}):
    if n in memo:  # if already calculated, return it
        return memo[n]
    if n <= 1:     # base cases
        return n
    memo[n] = fibo(n-1, memo) + fibo(n-2, memo)  # store in memo
    return memo[n]

print(fibo(5))  
print(fibo(10))

5
55


In [13]:
def fibo(n):
    a, b = 0, 1
    for i in range(n):
        a, b = b, a + b
    return a
fibo(10)

55

## 3) Create a function to return the longest substring without repeating characters.

In [18]:
def longest_substring(s):
    start = 0
    max_len = 0
    used_chars = {}
    
    for i, char in enumerate(s):
        if char in used_chars and start <= used_chars[char]:
            start = used_chars[char] + 1
        used_chars[char] = i
        max_len = max(max_len, i - start + 1)
    
    return max_len

print(longest_substring("abcabcbb"))  
print(longest_substring("bbbbb"))     
print(longest_substring("pwwkew"))    

3
1
3


In [26]:
n = input()
sub_str = ""      # current substring
max_len = 0

for i in n:
    if i in sub_str:
        sub_str = sub_str[sub_str.index(i) + 1:]  # remove repeated char and before it
    sub_str += i
    max_len = max(max_len, len(sub_str))

print(max_len)

 bb


1


## 4)  Write a function to determine if two strings are anagrams of each other.

In [27]:
def anagram(n1, n2):
    n1 = n1.replace(" ", "").lower()
    n2 = n2.replace(" ", "").lower()
    if sorted(n1) == sorted(n2):
        print("it is anagram")
    else:
        print("not anagram")
anagram("listen", "silent")

it is anagram


## 5)  Write a function to find the majority element in a list (element appearing > n/2 times).

In [35]:
def ele(n):
    for i in n:
        if n.count(i) > len(n) // 2:
            return i
    return None
print(ele([1, 2, 2, 2, 2, 2, 2, 2, 1]))

2


## 6) Implement a function that rotates a 2D matrix 90 degrees clockwise.

In [36]:
def rotate(matrix):
    n = len(matrix)
    res = [[0]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            res[j][n-1-i] = matrix[i][j]
    for row in res:
        print(row)

rotate([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

[7, 4, 1]
[8, 5, 2]
[9, 6, 3]


In [37]:
import numpy as np

def rotate_numpy(matrix):
    arr = np.array(matrix)
    rotated = np.rot90(arr, -1)  # -1 = clockwise
    print(rotated)

rotate_numpy([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

[[7 4 1]
 [8 5 2]
 [9 6 3]]


## 7) Write a function to find all prime numbers less than a given number n.

In [40]:
def prime(n):
    for i in range(2, n):
        is_prime = True
        for j in range(2, int(i**0.5) + 1):
            if i % j == 0:
                is_prime = False
                break
        if is_prime:
            print(i)

prime(21)

2
3
5
7
11
13
17
19


## 8) Create a function that takes a list of numbers and returns all the subsets.

In [ ]:
def lst(

In [7]:
n = [2, -3, -14, 7]
sum=0
for i in n:
    if i < 0:
        sum+=i
print(sum)

-17


In [8]:
n = list(map(int, input().split()))
total = 0
for i in n:
    if i < 0:
        total += i
print(total)

 2 -3 -14 7


-17


In [16]:
s = input()

word = ""
for ch in s:
    if ch.isupper() and word:  # uppercase starts a new word (but not for the first letter)
        print(word.swapcase()) # print previous word with swapped case
        word = ch
    else:
        word += ch

# print last word
if word:
    print(word.swapcase())

 saveChangesInTheEditor


SAVE
cHANGES
iN
tHE
eDITOR


In [17]:
s=input()
l=[0]
for i in range(len(s)):
    if(s[i].isupper()):
        l.append(i)
print(l)
for i in range(len(l)-1):
    print((s[l[i]:l[i+1]]).swapcase())
print((s[l[-1]:]).swapcase())

 saveChangesInTheEditor


[0, 4, 11, 13, 16]
SAVE
cHANGES
iN
tHE
eDITOR


## Take a list of space-separated integers and find the sum of all even numbers.

- Input:
- 5 8 3 10 7 2
- output
- 20

In [18]:
n = list(map(int, input().split()))
sum=0
for i in n:
    if i % 2==0:
        sum+=i
print(sum)   

 5 8 3 10 7 2


20


## Take a list of space-separated integers and count how many numbers are greater than 50.
- Input:

- 12 99 45 60 75 5

- Output:


- 3

In [19]:
n = list(map(int, input().split()))
count =0
for i in n:
    if i > 50:
        count+=1
print(count)

 12 99 45 60 75 5


3


## Given a list of space-separated integers, count how many numbers are positive and divisible by 3.

- Input:

- 3 6 -9 12 15 -18 20
- Output:
- 4


In [20]:
n = list(map(int, input().split()))
count =0
for i in n:
    if i >0 and i % 3==0:
        count+=1
print(count)

 3 6 -9 12 15 -18 20


4



## Given a list of space-separated integers, find the largest negative number. If there is no negative number, print 0.

5 -2 -7 10 4 -1
- Output:

- -1

In [23]:
n = list(map(int, input().split()))
max_negative = None
for i in n:
    if i<0:
        if max_negative is None or i> max_negative:
            max_negative = i
if max_negative is None:
    print(0)
else:
    print(max_negative)    

 5 -2 -7 10 4 -1


-1


In [29]:
n = int(input())
sum =0
for i in range(2, n+1, 3):
    for j in range(i, n, 2):
        sum = i + j
        if i > n:
            break
print(sum)

 123


244


In [32]:
n = input()
n=n.int()
for i in range(2, n+1):
    if i% 2!=0:
        sum =i
print(sum, end=",")  

 150


149,

In [33]:
n = int(input())
num = 2
diff = 3
while num <= n:
    print(num, end="")
    num += diff
    diff += 2
    if num <= n:
        print(",", end="")

 150


2,5,10,17,26,37,50,65,82,101,122,145

In [38]:
n = int(input())
num = 2
diff = 3

for i in range(n):  
    if num > n:
        break
    print(num, end="")
    num += diff
    diff += 2
    if num <= n:
        print(",", end="")

 150


2,5,10,17,26,37,50,65,82,101,122,145

### Question:
##### Print the sequence starting from 5, where each next number increases by an even number starting from 4. Stop when the number exceeds n (input from user). Separate numbers with commas.
- Input: 25  
- Output: 5,9,14,20


In [43]:
n = int(input())
num = 5
diff = 4

for i in range(n):
    if num > n:
        break
    print(num, end="")
    num += diff
    diff += 2
    if num <= n:
        print(",", end="")

 25


5,9,15,23

In [44]:
n = int(input())
num = 5
diff = 4

while num <= n:
    print(num, end="")
    num += diff
    diff += 2
    if num <= n:
        print(",", end="")

 25


5,9,15,23

In [46]:
n = int(input())
num = 5
diff = 4

for i in range(n):  
    if num > n:
        break
    print(num, end="")
    num += diff
    diff += 2
    if num <= n:
        print(",", end="")

 25


5,9,15,23

In [48]:
n = int(input())
num = 5
diff = 4

for _ in range(n):  
    if num > n:
        break
    print(num, end="")
    num += diff
    diff += 1
    if num <= n:
        print(",", end="")

 25


5,9,14,20

#### Input: 50
##### Output: 3,5,9,17,33


In [50]:
n = int(input())
num = 3
for i in range(n):
    if num> n:
        break
    print(num, end="")
    num = (num*2) - 1
    if num<=n:
        print(",", end="")

 50


3,5,9,17,33

## Sample Input
             9
             10 20 20 10 10 30 50 10 20
#### Sample Output
             3

###### Explanation
             Alex can match 3 pairs of socks i.e 10-10, 10-10, 20-20
             while the left out socks are 50, 60, 20

In [2]:
n = int(input())
n1 = list(map(int, input().split()))
pairs = 0
for i in set(n1):
    count = n1.count(i)
    pairs+=count//2
print(pairs)

 9
 10 20 20 10 10 30 50 10 20


3


In [12]:
n = input()
for i in range(len(n)-1):
    if abs(int(n[i]) - int(n[i+1])) != 1:
        print("Not a Step Number")
        break
else:
    print("Step Number")

 123


Step Number


#### Example Input & Output:

- Input 1:

- 18


- Output 1:

- Harshad Number

- ##### (Explanation: 1 + 8 = 9, and 18 is divisible by 9)

In [24]:
n = input()
for i in range(len(n)-1):
    digit = int(n[i]) + int(n[i+1])
    if int(n) % digit==0:
                print("Harshad number")
                break      
else:
    print("not")               

 19


not


In [31]:
n = input()
total = 0
for d in n:
    total += int(d)  # add each digit to total

if int(n) % total == 0:
    print("Harshad number")
else:
    print("Not a Harshad number")

 18


Harshad number


## Question:
- A number is called an Armstrong number if the sum of its digits raised to the power of the number of digits equals the number itself.

- Example:

- 153 → 1³ + 5³ + 3³ = 1 + 125 + 27 = 153 ✅ Armstrong number

- 9474 → 9⁴ + 4⁴ + 7⁴ + 4⁴ = 9474 ✅ Armstrong number

- 123 → 1³ + 2³ + 3³ = 36 ❌ Not Armstrong number

In [39]:
n = input()
total =0
for i in n:
    total+= int(i) ** len(n)
    
    if total == int(n):
               print("armstrong  number")
               break
else:
    print("not armstrong number")   

 153


armstrong  number


In [51]:
n = input()
reverse =""
for i in n:
    reverse = i + reverse
if n == reverse:
    print("Palidnrome")
else:
    print("Not")  

 sindhu


Not


In [48]:
n = input()
if n == n[::-1]:
    print("palindrome")
else:
    print("not palindrome")

 121


palindrome


## Question:
A number is called a Perfect number if the sum of its proper divisors (excluding itself) equals the number.

Example:

6 → divisors = 1, 2, 3 → sum = 1 + 2 + 3 = 6 ✅ Perfect number

28 → divisors = 1, 2, 4, 7, 14 → sum = 28 ✅ Perfect number

12 → divisors = 1, 2, 3, 4, 6 → sum = 16 ❌ Not perfect

In [56]:
n = int(input())
count = 0
for i in range(1, n):
    if n % i == 0:
        count+=i
if count == n:
    print("perfect number")
else:
    print("not a perfect number")    

 6


perfect number


## Question:
A number is called a Fibonacci number if it appears in the Fibonacci sequence.

The Fibonacci sequence starts like this: 0, 1, 1, 2, 3, 5, 8, 13, 21, …

Each number is the sum of the previous two numbers.

Task:
Write a program that:

Reads a number from input.

Checks whether the number appears in the Fibonacci sequence.

Prints "Fibonacci number" or "Not a Fibonacci number".

In [57]:
n = int(input())
a = 0
b =1
for i in range(0, n):
    a, b+= b, a
    if n == a:
        print("yes")
        break
else:
    print("no")

SyntaxError: 'tuple' is an illegal expression for augmented assignment (2198802835.py, line 5)

## 23) Count frequency of characters in a string.
# Input: "programming"
# Output: {'p':1, 'r':2, 'o':1, 'g':2, 'a':1, 'm':2, 'i':1, 'n':1}
# Explanation: Count of each character is calculated.


In [4]:
n = input()
freq={}
for i in n:
    freq[i] = freq.get(i, 0) + 1   
print(freq)

 programming


{'p': 1, 'r': 2, 'o': 1, 'g': 2, 'a': 1, 'm': 2, 'i': 1, 'n': 1}


### i = 'p'

freq.get('p', 0) → 'p' not in dictionary → returns 0

0 + 1 = 1

Store this in dictionary: freq['p'] = 1

### 30) Find common elements in two lists.
### Input: [1, 2, 3, 4], [3, 4, 5, 6]
#### Output: [3, 4]
### Explanation: 3 and 4 appear in both lists.

In [7]:
a1 = eval(input())
a2= eval(input())
dig =[]
for i in a1:
    if i in a2:
        dig.append(i)
print(dig)     

 [1, 2, 3, 4]
 [3, 4, 5, 6]


[3, 4]


In [8]:
a1 = eval(input())
a2= eval(input())
dig = []
for i in a1:
    if i in a2 and i not in dig:
        dig.append(i)
print(dig)  

 [1, 2, 3, 4]
 [3, 4, 5, 6]


[3, 4]


In [ ]:
a1 = eval(input())
a2 = eval(input())
          

In [2]:
n = int(input())
if n % 2==0:
    print(n, " is even")
else:
    print(n, " is not even")

 3


3  is not even


In [4]:
n = list(map(int, input().split()))
max1= n[0]
for i in n:
    if i>max1:
        max1= i
print(max1)

 10 25 15


25


In [1]:
x = [0] * 3
x[0] = 5
print(x)

[5, 0, 0]


In [3]:
for i in range(1, 5, 3):
    print(i, end=" ")

1 4 

In [5]:
x = (1, 2, [3, 4])
x[2].append(5)
print(x)

(1, 2, [3, 4, 5])


In [5]:
n = 10
for i in range(n, 0, -1):
    print(i)
    

10
9
8
7
6
5
4
3
2
1


In [1]:
arr = [1, 2, 2, 3, 1]

freq = {}

for x in arr:
    if x in freq:
        freq[x] += 1
    else:
        freq[x] = 1

print(freq)


{1: 2, 2: 2, 3: 1}


In [2]:
d=int(input())
t=int(input())
if(d<0 or t<1 or t>60):
  print("error")
s=int((d/t)*18/5)
print(s)

 500
 30


60
